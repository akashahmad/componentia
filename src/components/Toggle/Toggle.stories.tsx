import * as React from "react";

import type { Meta, StoryObj } from "@storybook/react";

import { Toggle } from "./Toggle";

const meta: Meta<typeof Toggle> = {
    component: Toggle,
    tags: ["autodocs"],
};

export default meta;
type Story = StoryObj<typeof Toggle>;

export const Warning: Story = {
    args: {
        defaultState: false,
        state: false,
    },
    argTypes: {
        setState: {
            control: false,
            table: {
                defaultValue: { summary: "undefined" },
            },
        },
    },
    render: props => {
        const [checked, setChecked] = React.useState(props.state!);

        return <Toggle state={checked} setState={setChecked} {...props} />;
    },
};
