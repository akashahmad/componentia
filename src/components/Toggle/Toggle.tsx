import * as React from "react";

import cx from "classnames";

import { ToggleCheckedIcon } from "../icons/ToggleCheckedIcon";
import { ToggleCircularActiveIcon } from "../icons/ToggleCircularActiveIcon";
import { ToggleCircularUnActiveIcon } from "../icons/ToggleCircularUnActiveIcon";
import { ToggleCrossIcon } from "../icons/ToggleCrossIcon";

export function Toggle({ setState, defaultState = false, variant = "default", state = false }: ToggleProps) {
    const isMounted = React.useRef(false);
    const [checked, setChecked] = React.useState(() => defaultState);

    const handleToggle = () => {
        setChecked(prev => !prev);
    };

    React.useEffect(() => {
        setChecked(state);
        setState && setState(state);
    }, [state]);

    React.useEffect(() => {
        if (isMounted.current) {
            setState && setState(checked);
        } else {
            setChecked(state || defaultState);
        }

        isMounted.current = true;
    }, [checked]);

    const Icon =
        variant === "default" ? (
            <span
                className={cx("w-7 h-7 rounded-full inline-block absolute top-1/2 translate-y-[-50%]", {
                    ["right-[2px] bg-white"]: checked,
                    ["left-[2px] bg-white"]: !checked,
                })}
            />
        ) : variant === "checkmark" ? (
            <span
                className={cx("rounded-full absolute top-1/2 translate-y-[-50%] flex items-center justify-start", {
                    ["right-[-2px]"]: checked,
                    ["left-[-2px]"]: !checked,
                })}
            >
                {checked ? <ToggleCheckedIcon /> : <ToggleCrossIcon />}
            </span>
        ) : (
            variant === "mini-circle" && (
                <span
                    className={cx("rounded-full absolute top-1/2 translate-y-[-50%] flex items-center justify-start", {
                        ["right-[-2px]"]: checked,
                        ["left-[-2px]"]: !checked,
                    })}
                >
                    {checked ? <ToggleCircularActiveIcon /> : <ToggleCircularUnActiveIcon />}
                </span>
            )
        );

    return (
        <div
            onClick={handleToggle}
            className={cx("h-8 w-[55px] bg-dark rounded-full relative cursor-pointer transition-all select-none", {
                ["bg-primary"]: checked && variant === "default",
                ["bg-grey-3 "]: checked && variant !== "default",
                ["bg-grey-8"]: variant === "mini-circle",
                ["bg-grey-3"]: !checked,
            })}
        >
            {Icon}
        </div>
    );
}

interface ToggleProps {
    defaultState?: boolean;
    setState?: React.Dispatch<React.SetStateAction<boolean>>;
    variant?: "default" | "checkmark" | "mini-circle";
    state?: boolean;
}
