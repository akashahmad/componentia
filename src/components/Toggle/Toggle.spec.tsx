import * as React from "react";

import { render } from "@testing-library/react";
import { Toggle } from "./Toggle";

test("Toggle component is rendering properly", async () => {
    const { container } = render(<Toggle />);

    expect(container).toMatchSnapshot();
});
