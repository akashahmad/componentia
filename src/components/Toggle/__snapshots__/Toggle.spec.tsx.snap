// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle component is rendering properly 1`] = `
<div>
  <div
    class="h-8 w-[55px] bg-dark rounded-full relative cursor-pointer transition-all select-none bg-grey-3"
  >
    <span
      class="w-7 h-7 rounded-full inline-block absolute top-1/2 translate-y-[-50%] left-[2px] bg-white"
    />
  </div>
</div>
`;
